cpp/src/parquet/encoding-test.cc

```diff
@@ -314,6 +318,185 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
   ASSERT_THROW(MakeDictDecoder<BooleanType>(nullptr), ParquetException);
 }
 
+// ----------------------------------------------------------------------
+// Shared arrow builder decode tests
+class TestDecodeArrow : public ::testing::Test {
+ public:
+  void SetUp() override {
+    InitDataInputs();
+    SetupEncoderDecoder();
+  }
+
+  void InitDataInputs() {
+    // Build a dictionary array and its dense representation for testing building both
+    auto dict_values = ::arrow::ArrayFromJSON(::arrow::binary(), "[\"foo\", \"bar\"]");
+    auto dtype = ::arrow::dictionary(::arrow::int8(), dict_values);
+    auto indices = ::arrow::ArrayFromJSON(::arrow::int8(), "[0, 1, 0, 0]");
+    ASSERT_OK(::arrow::DictionaryArray::FromArrays(dtype, indices, &expected_dict_));
+    expected_array_ =
+        ArrayFromJSON(::arrow::binary(), "[\"foo\", \"bar\", \"foo\", \"foo\"]");
+    num_values_ = static_cast<int>(expected_array_->length());
+
+    // arrow::BinaryType maps to parquet::ByteArray
+    const auto& binary_array = static_cast<const ::arrow::BinaryArray&>(*expected_array_);
+    for (int64_t i = 0; i < binary_array.length(); i++) {
+      auto view = binary_array.GetView(i);
+      input_data_.emplace_back(static_cast<uint32_t>(view.length()),
+                               reinterpret_cast<const uint8_t*>(view.data()));
+    }
+
+    valid_bits_ = vector<uint8_t>(::arrow::BitUtil::BytesForBits(num_values_) + 1, 255);
+  }
+
+  // Setup encoder/decoder pair for testing with
+  virtual void SetupEncoderDecoder() = 0;
+
+  void CheckDecodeDense(const int actual_num_values,
+                        ::arrow::internal::ChunkedBinaryBuilder& builder) {
+    ASSERT_EQ(actual_num_values, num_values_);
+    ::arrow::ArrayVector actual_vec;
+    ASSERT_OK(builder.Finish(&actual_vec));
+    ASSERT_EQ(actual_vec.size(), 1);
+    ASSERT_TRUE(actual_vec[0]->Equals(expected_array_))
```

Comment:
Can you use ASSERT_ARRAYS_EQUAL here?

Reply:
Good tip!  I didn't know about `ASSERT_ARRAYS_EQUAL`.